pkg/hubble/recorder/sink/dispatch.go

```diff
@@ -0,0 +1,231 @@
+}
+
+// RegisterSink registers a new sink for the given rule ID. Any captures with a
+// matching rule ID will be forwarded to the pcap sink w. The provided header
+// is written to the pcap sink w upon initialization.
+func (d *Dispatch) RegisterSink(ctx context.Context, ruleID uint16, w pcap.RecordWriter, header pcap.Header) (*Handle, error) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	if _, ok := d.sinkByRuleID[ruleID]; ok {
+		return nil, fmt.Errorf("sink for rule id %d already registered", ruleID)
+	}
+
+	s := startSink(ctx, w, header, d.sinkQueueSize)
+	d.sinkByRuleID[ruleID] = s
+	return &Handle{
+		C:    s.trigger,
+		sink: s,
+	}, nil
+}
+
+// UnregisterSink will stop and unregister the sink for the given ruleID.
+// It waits for any pending packets to be forwarded to the sink before closing
+// it and returns the final statistics or an error, if an error occurred.
+func (d *Dispatch) UnregisterSink(ctx context.Context, ruleID uint16) (stats Statistics, err error) {
+	d.mutex.Lock()
+	s, ok := d.sinkByRuleID[ruleID]
+	delete(d.sinkByRuleID, ruleID)
+	// unlock early to avoid holding the lock during s.close() which may block
+	d.mutex.Unlock()
+
+	if !ok {
+		return Statistics{}, fmt.Errorf("no sink found for rule id %d", ruleID)
+	}
+
+	if err = s.close(ctx); err != nil {
+		return Statistics{}, err
+	}
+
+	return s.copyStats(), nil
+}
+
+func (d *Dispatch) decodeRecordCaptureLocked(data []byte) (rec record, err error) {
+	dataLen := uint32(len(data))
+	if dataLen < monitor.RecorderCaptureLen {
+		return record{}, fmt.Errorf("not enough data to decode capture message: %d", dataLen)
+	}
+
+	// This needs to stay in sync with struct capture_msg from
+	// bpf/include/pcap.h.
+	// We could use binary.Read on monitor.RecorderCapture, but since it
+	// requires reflection, it is too slow to use on the critical path here.
+	const (
+		offsetRuleID         = 2
+		offsetBootTime       = 8
+		offsetCaptureLength  = 16
+		offsetOriginalLength = 20
+	)
+	n := byteorder.Native
+	ruleID := n.Uint16(data[offsetRuleID:])
+	bootTime := n.Uint64(data[offsetBootTime:])
+	capLen := n.Uint32(data[offsetCaptureLength:])
+	origLen := n.Uint32(data[offsetOriginalLength:])
+
+	// data may contain trailing garbage from the perf ring buffer
+	// https://lore.kernel.org/patchwork/patch/1244339/
+	packetEnd := monitor.RecorderCaptureLen + capLen
+	if dataLen < packetEnd {
+		return record{}, fmt.Errorf("capture record too short: want:%d < got:%d", dataLen, packetEnd)
+	}
+	packet := data[monitor.RecorderCaptureLen:packetEnd]
+
+	return record{
+		timestamp: d.bootTimeToWallTimeLocked(int64(bootTime)),
+		ruleID:    ruleID,
+		inclLen:   capLen,
+		origLen:   origLen,
+		data:      packet,
+	}, nil
+}
+
+func (d *Dispatch) bootTimeToWallTimeLocked(bootTime int64) time.Time {
+	elapsedSinceStart := time.Duration(bootTime - d.startBootTime)
+	return d.startWallTime.Add(elapsedSinceStart)
+}
+
+func getTimeNow() (bootTime int64, wallTime time.Time, err error) {
```

Comment:
I've implemented something similar here: https://github.com/ti-mo/conntracct/blob/master/pkg/boottime/boottime.go. I opted to expose `runtime.nanotime()` in my package so I wouldn't have to make a slow syscall, and vDSO is used as a source for both timestamps. Feel free to take a look, it's well-documented.

A few traps I've fallen into while implementing this:
- There can be an arbitrary amount of time between the execution of `time.Now()` and `unix.ClockGettime()`, especially under scheduler pressure during program startup. This can be slightly improved by running these 2 calls under `runtime.LockOSThread()`, which at least takes the Go scheduler out of the equation. Ideally, these results need to be sampled and averaged to eliminate (OS) scheduler jitter. (and it still won't be accurate :sweat_smile:)
- Hardware timer pauses (during machine/laptop suspend, VM migrations or other virtualization-related artifacts) invalidate this offset. `CLOCK_BOOTTIME` does not advance during a hardware pause, so it's possible for 2 events that occurred right before and right after a (let's say, 1hr) pause window to have timestamps that are 1ns apart. This happens much more often than you'd expect. (the cloud is someone else's computer after all) This will result in events seemingly occurring earlier than they actually have, causing events to fall outside of a pcap window. (being a few seconds off can already be an issue if a user runs e.g. a `curl` right after starting a pcap) The solution I came up with is refreshing the offset every 5/10 seconds or so, since userspace has no way of knowing when these pauses occur.

Hope this helps!

Reply:
Thanks a ton for the in-depth reply here. I haven't considered the scheduler issue, I might add a `LockOSThread`, but yeah, anything we do here will always be best-effort I'm afraid.

So the reason I picked `CLOCK_BOOTTIME`, is because this is what the datapath is using, see:
https://github.com/cilium/cilium/blob/2ff996c21ad83fb0ca21ce06ea10d45498825f65/bpf/lib/pcap.h#L78-L83
`bpf_ktime_get_boot_ns` is CLOCK_BOOTTIME, see https://lkml.org/lkml/2020/4/20/1443 

If `CLOCK_BOOTTIME` is not what we want, then we need to fix that in the datapath first. **But**, I actually believe we _do_ want `CLOCK_BOOTTIME`. My reading is that `CLOCK_BOOTTIME` is _not_ affected by your concerns regarding suspension (while `runtime.nanotime()`, which is `CLOCK_MONOTONIC`, _is_ affected)

> CLOCK_BOOTTIME (since Linux 2.6.39; Linux-specific)
    Identical to CLOCK_MONOTONIC, except it also includes any time that the system is suspended. This allows applications to get a suspend-aware monotonic clock without having to deal with the complications of CLOCK_REALTIME, which may have discontinuities if the time is changed using settimeofday(2). 
https://linux.die.net/man/2/clock_gettime

Unless I'm misinterpreting something here I'll need to stick with `CLOCK_BOOTTIME` which `runtime` unfortunately does not expose (see also https://github.com/golang/go/issues/24595). But I like your linker trick to access it anyway :D

Reply:
> bpf_ktime_get_boot_ns is CLOCK_BOOTTIME, see https://lkml.org/lkml/2020/4/20/1443

Oh, wasn't aware `ktime_get_boot_ns` made it in, my implementation predates that helper, and so relies on monotonic being available only. 100% correct on the `_MONOTONIC` vs. `_BOOTTIME`, didn't catch that.

For posterity, the kernel commit that introduced it (71d19214776e) only landed in Linux 5.8, so this lb-only datapath code is not backwards-compatible.  `bpf_ktime_cache_set(boot_ns)` does indeed call `ktime_get_boot_ns()`, so the 5.8 (or backport kernel) req should probably be documented in https://github.com/cilium/cilium/pull/15633.

@borkmann for future (and backwards) compat, userspace is going to need to know what the BPF event's clock source is (mono/boottime/timeofday, mentioning all three since we might want to target pre-5.8 at some point). Maybe we should already extend `struct pcap_timeoff` and make sure userspace can make that distinction? (thinking about upgrade safety)

Reply:
Right, the https://github.com/cilium/cilium/commit/02e55a792325f82efbaf95b6bac7de1dd8e21f2c will check for availability and bail out if `--enable-recorder=true` is used and kernel doesn't have the boot time source. Potentially if users don't care too much about the specific clock, we could also use monotonic and don't translate anything from hubble side (apart from timeval conversion). I would probably wait till such need comes up. After my blocker PRs are done, I'm planning to document all stand-alone LB features as a getting started guide for Cilium in detail, so I'll definitely include such kernel requirements there (right now only the agent will tell it in its error, but a GSG of similar scope and depth as https://docs.cilium.io/en/v1.9/gettingstarted/kubeproxy-free/ would be super useful for the LB-only mode).

Reply:
I have updated the mechanism to fetch the diff multiple times and take the minimum. I also simplified the clock offset calculation a bit.